Support Vec, HashMap and more types

sw1sh opened the pull request:
I'm a Rust noob, so some things could probably be improved. I'm trying to build a paclet for this small library: https://github.com/jcmgray/cotengrust/blob/main/src/lib.rs. With these additional types and traits, I've been able to just use #[wll::export] attribute on a function I need:

```rust
#[wll::export]
fn optimize_optimal(
    inputs: Vec<Vec<char>>,
    output: Vec<char>,
    size_dict: Dict<char, f32>,
    minimize: Option<String>,
    cost_cap: Option<Score>,
    search_outer: Option<bool>,
    simplify: Option<bool>,
    use_ssa: Option<bool>,
) -> Vec<Vec<Node>>
```
And it works without fiddling with wstp. Pretty cool stuff! :)

sw1sh committed:
Support Vec and HashMap with DataStore

ConnorGray commented on wolfram-library-link/src/args.rs:
```diff
@@ -254,6 +256,16 @@ impl<'a> FromArg<'a> for &'a str {
+impl FromArg<'_> for char {
+    unsafe fn from_arg(arg: &MArgument) -> Self {
+        String::from_arg(arg).chars().next().unwrap()
+    }
```
Hi @sw1sh, as discussed in our meeting, the basic implementation here is exactly right, but we may want to improve the error reporting to help users out in cases where they accidentally pass in strings that are not a single `char` in length.

My proposal would be something like:
```rust

unsafe fn from_arg(arg: &MArgument) -> Self {
    let string = String::from_arg(arg);

    let mut chars = string.chars();

    let Some(char) = chars.next() else {
        panic!("char argument string was empty")
    };

    if chars.next().is_some() {
        panic!("char argument string contains multiple chars: {string}")
    }

    char
}
```

sw1sh committed:
Add u64 DataStoreAdd impl and u64 DataStoreNodeValue conversion tests

sw1sh committed:
Adds DataStore support for `usize` type

Extends the `DataStore` to support the `usize` type by implementing the `DataStoreAdd` trait.

This enables storing `usize` values in the `DataStore`, handling potential overflow by ensuring that the value does not exceed `i64::MAX`.

Also enables `non_snake_case` linting for the `rtl.rs` module to mirror Wolfram RTL field names, improving consistency with the C API.

sw1sh committed:
Adds From impls for u32 and u16

Implements `From>` for `u32` and `u16`.

This allows for direct conversion of a `DataStoreNodeValue::Integer`
into these unsigned integer types. This simplifies data handling within
the Wolfram Library Link.

sw1sh committed:
Updates panic hook and numeric array kind

Updates the panic hook to use `PanicHookInfo` for compatibility with newer Rust versions.

Changes `NumericArray::kind` to return `NumericArrayKind` to clarify lifetime management and prevent potential dangling references.

Adds a conditional check to prevent panics when attempting to store `usize` values that exceed `i64::MAX` in the DataStore.

sw1sh committed:
Updates LibraryLink bindings generation

Updates the target Wolfram version to 14.3.0 and includes corresponding bindings.

Improves the error message displayed when pre-generated bindings
are missing, now including a listing of available versions to assist
users in selecting or generating the correct bindings.

Provides guidance on generating bindings via a link to Maintenance.md.

sw1sh committed:
Adds u8, u16, and u32 support to DataStore

Extends the `DataStore` functionality to support `u8`, `u16`, and `u32` types.

This change introduces implementations of the `DataStoreAdd` trait for these unsigned integer types, allowing them to be seamlessly added to a `DataStore`. Corresponding test cases and Wolfram Language bindings are included to verify the correctness of these new features.

sw1sh committed:
Adds support for i8, i16, and i32 types in DataStore

Extends the DataStore functionality to include support for i8, i16,
and i32 integer types.

This change adds the necessary `DataStoreAdd` implementations and
`From` conversions, as well as new functions to
round-trip and retrieve these types from the DataStore.  It also includes
corresponding tests to verify the new functionality.

sw1sh committed:
scripts: add Docker helper to generate LibraryLink bindings using WolframEngine image; update docs

sw1sh committed:
Adds Dockerfile for binding generation

Adds a Dockerfile to streamline the generation of Wolfram LibraryLink Rust bindings.

This provides a reproducible environment, encapsulating the necessary dependencies
(WolframEngine, Rust toolchain, build tools) to generate bindings without
requiring manual setup on the host system.

Also includes a script to manage the binding generation process and checks for
new, changed, or removed files, exiting with an error if no new or changed
bindings are produced.

sw1sh committed:
Adds support for u32, u64, usize, isize

Adds `FromArg` and `IntoArg` implementations for the `u32`, `u64`, `usize`, and `isize` types, allowing them to be used as native Wolfram LibraryLink arguments.

Also includes corresponding Wolfram Language tests to verify the proper round-trip conversion of these types.

sw1sh committed:
Adds support for BigInt and BigUint types

Adds support for `num_bigint::BigInt` and `num_bigint::BigUint` types as arguments and return values for Wolfram LibraryLink functions.

This change allows Wolfram Language code to seamlessly interact with Rust functions that utilize arbitrary-precision integers. It includes necessary type conversions and tests to ensure correct data handling.

sw1sh committed:
Implements `IntoArg` for tuples

Implements the `IntoArg` trait for tuples of varying lengths.

This allows tuples to be passed as arguments to Wolfram LibraryLink
functions, where they are converted to a `DataStore`.

sw1sh committed:
Implements `DataStoreAdd` for tuples

Implements the `DataStoreAdd` trait for tuples of varying sizes,
allowing tuples to be directly added to a `DataStore`.

This simplifies the process of storing multiple values in a
`DataStore` by treating tuples as single units.

sw1sh committed:
Adds tuple conversion from DataStoreNodeValue

Introduces a macro to implement conversion from DataStoreNodeValue to tuples of up to 12 elements.

riccardodivirgilio commented:
This repository is being archived. Active development of the `wolfram-library-link` crate has moved to the [`wolfram-rust-library`](https://github.com/WolframResearch/wolfram-rust-library) monorepo, in the [`wolfram-library-link/`](https://github.com/WolframResearch/wolfram-rust-library/tree/HEAD/wolfram-library-link) directory.

GitHub does not allow transferring pull requests between repositories, so we are closing this PR as part of the archive. **Please re-open your changes against the monorepo**: https://github.com/WolframResearch/wolfram-rust-library

If your change is tied to an issue, you can find the migrated issue list here: https://github.com/WolframResearch/wolfram-rust-library/issues?q=is%3Aissue+label%3Acrate%3Awolfram-library-link

Thanks for your contribution!
